### PR TITLE
Generate CSV file from all test evidences for ledger client components

### LINKED ISSUE #15223: Provide a CSV file for the security evidence of runtime components
Make the information available together with the one provided by the Canton and Ledgers team as a CSV file bundled with every release, to which the documentation points

Currently, the `TEST_EVIDENCE` annotations are processed to generate the `security-evidence.md` file.

Either adapt the generator to (additionally) output a CSV file or move to scalatest style annotations (lib-scala/test-evidence).

Part of #14254

### PR DESCRIPTION
**Pull request by cbley-da**
### Description

This PR replaces the TEST_EVIDENCE annotations which are processed by the `security:evidence-security` tool with the test-evidence framework in order to generate CSV and JSON entries for them.

Fixes #15223

_Note_: All commits starting with "REMOVE" are only here to help reviewing the changes; I will remove them once approved. More specifically, that means :

* there will be no changes to CI files
* the `daml-security-test-evidence.csv` will disappear

### Follow Up

1. deprecate old categories in libs-scala/test-evidence/tag/src/main/scala/com/daml/test/evidence/tag/Security.scala
2. refine _asset_, _happyPath_ or _attack_ property of SecurityTest entries (see https://github.com/digital-asset/daml/issues/15699)

### CODE REVIEW

**Comment by matthiasS-da** [comment]
Later today, I would like to have a look at the test annotations.

File: ledger-service/http-json/src/failurelib/scala/http/FailureTests.scala
Comment (matthiasS-da) [inline ledger-service/http-json/src/failurelib/scala/http/FailureTests.scala:55]: Please make sure that the annotations include a scenario (either happy case  or attack). It is important information to understand what the test is actually about.

File: ledger-service/http-json/src/failurelib/scala/http/FailureTests.scala
Comment (matthiasS-da) [inline ledger-service/http-json/src/failurelib/scala/http/FailureTests.scala:85]: Given that you are using toxiproxy to inject failures on the connection, I would expect this is more of a ReliabilityTest than a security test.

File: ledger-service/http-json/src/it/scala/http/AuthorizationTest.scala
Comment (matthiasS-da) [inline ledger-service/http-json/src/it/scala/http/AuthorizationTest.scala:83]: This seems to drop contents from the `TEST_EVIDENCE` comment, namely:
```
Authorization: Updating the package service fails with insufficient authorization
```

Feel free to rewrite / enhance information, but do not drop information.

**Comment by matthiasS-da** [comment]
I think we should have a call to talk this through.

File: ledger-service/http-json/BUILD.bazel
Comment (S11001001) [inline ledger-service/http-json/BUILD.bazel:299]: You can remove this; these aliases are only there for people who were targeting the pre-split names, and naturally there was never a `tests-lib` before.

File: ledger-service/http-json/BUILD.bazel
Comment (S11001001) [inline ledger-service/http-json/BUILD.bazel:227]: Do the tests still run when you run this target?

File: ledger-service/http-json/BUILD.bazel
Comment (S11001001) [inline ledger-service/http-json/BUILD.bazel:628]: Are we now compiling the integration tests 4 times instead of 2?

If so, shouldn't this dep be removed?

If not, do the tests run with this arrangement? (Since the `_suite` variant behaves quite differently from the ordinary variant.)

File: ledger-service/http-json/BUILD.bazel
Comment (S11001001) [inline ledger-service/http-json/BUILD.bazel:754]: Please remove instead of commenting out.

File: ledger-service/http-json/BUILD.bazel
Comment (S11001001) [inline ledger-service/http-json/BUILD.bazel:820]: Same question here, as this is a completely different strategy from what you're doing with the integration tests.

File: libs-scala/test-evidence/generator/src/main/scala/org/scalatest/ScalaTestAdapter.scala
Comment (S11001001) [inline libs-scala/test-evidence/generator/src/main/scala/org/scalatest/ScalaTestAdapter.scala:8]:
```suggestion
```

File: triggers/service/BUILD.bazel
Comment (S11001001) [inline triggers/service/BUILD.bazel:266]: Another case of doubling compilation.

File: triggers/service/auth/BUILD.bazel
Comment (S11001001) [inline triggers/service/auth/BUILD.bazel:176]: Also doubling compilation.

File: triggers/service/src/test/scala/com/digitalasset/daml/lf/engine/trigger/TriggerServiceTest.scala
Comment (S11001001) [inline triggers/service/src/test/scala/com/digitalasset/daml/lf/engine/trigger/TriggerServiceTest.scala:58]: Please put this locally in the smallest scope that needs it.

**Comment by matthiasS-da** [comment]
Thanks for the productive zoom meeting!
Let's do it as follows:
- Please have another run on the test tags, as discussed. Ping me through slack once you have done that and I will review the test tags.
- I don't feel confident to review the bazel changes. Please find another reviewer. Up to you whether you split it into separate PRs vs. do all the changes together in this PR.

File: ledger-service/http-json/BUILD.bazel
Comment (cbley-da) [reply ledger-service/http-json/BUILD.bazel:227]: Nope. Looking into it...

File: triggers/service/src/test/scala/com/digitalasset/daml/lf/engine/trigger/TriggerServiceTest.scala
Comment (cbley-da) [reply triggers/service/src/test/scala/com/digitalasset/daml/lf/engine/trigger/TriggerServiceTest.scala]: Done.

File: ledger-service/http-json/BUILD.bazel
Comment (cbley-da) [reply ledger-service/http-json/BUILD.bazel]: Done.

File: ledger-service/http-json/src/failurelib/scala/http/FailureTests.scala
Comment (cbley-da) [reply ledger-service/http-json/src/failurelib/scala/http/FailureTests.scala:55]: This PR is mostly a mechanical transformation from TEST_EVIDENCE comments to the test-evidence scalatest tags.

As agreed in our call, I'll take care of this in another PR.

File: ledger-service/http-json/BUILD.bazel
Comment (cbley-da) [reply ledger-service/http-json/BUILD.bazel]: Done.

File: ledger-service/http-json/BUILD.bazel
Comment (cbley-da) [reply ledger-service/http-json/BUILD.bazel]: I have reworked the PR to address this issues. But I am wondering, is there a simple way to see how many tests where run on CI? In order to compare between main and this branch?

**Comment by ray-roestenburg-da** [comment]
CI is failing @cbley-da

**Comment by cbley-da** [comment]
> CI is failing @cbley-da

Yes, I am looking into it...

File: triggers/service/auth/src/test/scala/com/daml/auth/middleware/oauth2/TestMiddleware.scala
Comment (remyhaemmerle-da) [inline triggers/service/auth/src/test/scala/com/daml/auth/middleware/oauth2/TestMiddleware.scala:55]: Can you link an issue or at least add a comment about this TDB. I have no idea what this is about.

File: daml-security-test-evidence.csv
Comment (pbatko-da) [inline daml-security-test-evidence.csv:1]: Could this file be generated into a subdirectory (e.g. 'generated/') instead of the repository's root?

Comment (cbley-da) [reply daml-security-test-evidence.csv:1]: This file is not to be kept. It was / is just there to be able to inspect the generated CSV file.

Comment (cbley-da) [reply daml-security-test-evidence.csv]: I rebased and removed all commits labeled with "REMOVE" now.

File: ledger-service/http-json/src/failurelib/scala/http/FailureTests.scala
Comment (cbley-da) [reply ledger-service/http-json/src/failurelib/scala/http/FailureTests.scala:85]: Yes, I think you are right. But since it was tagged as a Availability test before I rather would like to create a separate issue for it.

File: ledger-service/http-json/src/test/scala/com/digitalasset/http/CliSpec.scala
Comment (matthiasS-da) [inline ledger-service/http-json/src/test/scala/com/digitalasset/http/CliSpec.scala:279]: I prefer the previous test evidence description `TLS configuration is parsed correctly from the config file`. Could you please take that?

File: libs-scala/test-evidence/tag/src/main/scala/com/daml/test/evidence/tag/Security.scala
Comment (matthiasS-da) [inline libs-scala/test-evidence/tag/src/main/scala/com/daml/test/evidence/tag/Security.scala:116]: Could you please:
- Avoid duplicate properties (Authenticity vs Authentication, Privacy vs Confidentiality). To avoid migration effort, I would keep the old categories and add comment of the form `This is also known as Authentication.`.
- Add a ScalaDoc to new properties to explain what this is about. In particular for Resilience (against what?).

File: test-evidence/src/main/scala/com/daml/test/evidence/generator/Main.scala
Comment (matthiasS-da) [inline test-evidence/src/main/scala/com/daml/test/evidence/generator/Main.scala:35]: In the `println` statement, mention `file` instead of `fileName`, for improved clarity.

File: triggers/service/auth/src/test/scala/com/daml/auth/middleware/oauth2/TestMiddleware.scala
Comment (matthiasS-da) [inline triggers/service/auth/src/test/scala/com/daml/auth/middleware/oauth2/TestMiddleware.scala:402]: Please extend: `the /login endpoint with an oauth server checking claims`

File: triggers/service/auth/src/test/scala/com/daml/auth/middleware/oauth2/TestMiddleware.scala
Comment (matthiasS-da) [inline triggers/service/auth/src/test/scala/com/daml/auth/middleware/oauth2/TestMiddleware.scala:435]: Please extend: `the /login endpoint with an oauth server checking claims`

File: triggers/service/auth/src/test/scala/com/daml/auth/middleware/oauth2/TestMiddleware.scala
Comment (matthiasS-da) [inline triggers/service/auth/src/test/scala/com/daml/auth/middleware/oauth2/TestMiddleware.scala:487]: the /login endpoint with an oauth server checking claims

File: ledger-service/http-json/src/failurelib/scala/http/FailureTests.scala
Comment (cbley-da) [reply ledger-service/http-json/src/failurelib/scala/http/FailureTests.scala:85]: Note, I created a new issue for this: https://github.com/digital-asset/daml/issues/15714

File: triggers/service/auth/src/test/scala/com/daml/auth/middleware/oauth2/TestMiddleware.scala
Comment (matthiasS-da) [inline triggers/service/auth/src/test/scala/com/daml/auth/middleware/oauth2/TestMiddleware.scala:550]: `the TestMiddlewareClientNoRedirectToLogin client`

It's important to point out the client used, because otherwise we get:
- Test case 1: should not redirect
- Test case 2: should redirect
which sounds contradictory to the uninitiated observer.

File: triggers/service/auth/src/test/scala/com/daml/auth/middleware/oauth2/TestMiddleware.scala
Comment (matthiasS-da) [inline triggers/service/auth/src/test/scala/com/daml/auth/middleware/oauth2/TestMiddleware.scala:601]: the TestMiddlewareClientYesRedirectToLogin client

File: triggers/service/auth/src/test/scala/com/daml/auth/middleware/oauth2/TestMiddleware.scala
Comment (matthiasS-da) [inline triggers/service/auth/src/test/scala/com/daml/auth/middleware/oauth2/TestMiddleware.scala:637]: the TestMiddlewareClientAutoRedirectToLogin client

**Review by matthiasS-da — APPROVED** [review]
Thanks for cleaning this up!
My most critical comment is about not duplicating test properties (Authentication / Authenticity). Otherwise, I left a few comments of a more MINOR nature. Please address them before merging.

Pre-approving, assuming that you won't merge any duplicate test properties.

File: test-evidence/src/main/scala/com/daml/test/evidence/generator/Main.scala
Comment (cbley-da) [reply test-evidence/src/main/scala/com/daml/test/evidence/generator/Main.scala:51]: Yes, but I did not like this because of the literal numbers inside the different branches...
```scala
val exitCode = try {
  val securityTestEntries = TestEntryLookup.securityTestEntries
  val csvEntries = securityTestEntries.map(SecurityTestEntryCsv.apply)
  val csvFileName = args(0)
  val jsonFileName = args(1)

  writeEvidenceToCsvFile(csvFileName, csvEntries)
  writeEvidenceToJsonFile(jsonFileName, securityTestEntries)

  0
} catch {
  case NonFatal(e) =>
    e.printStackTrace()
    1
}

// explicitly exit, since some test suites have started non-daemon threads that prevent JVM shutdown
sys.exit(exitCode)
```
Alternatively, calling sys.exit twice:
```scala
try {
  val securityTestEntries = TestEntryLookup.securityTestEntries
  val csvEntries = securityTestEntries.map(SecurityTestEntryCsv.apply)
  val csvFileName = args(0)
  val jsonFileName = args(1)

  writeEvidenceToCsvFile(csvFileName, csvEntries)
  writeEvidenceToJsonFile(jsonFileName, securityTestEntries)

  // explicitly exit, since some test suites have started non-daemon threads that prevent JVM shutdown
  sys.exit(1)
} catch {
  case NonFatal(e) =>
    e.printStackTrace()
    // explicitly exit, since some test suites have started non-daemon threads that prevent JVM shutdown
    sys.exit(1)
}
```
WDYT?

Comment (matthiasS-da) [reply test-evidence/src/main/scala/com/daml/test/evidence/generator/Main.scala:51]: The literal numbers are pretty standard in scala (in java you would need to add `return`). 

Calling `sys.exit` twice is viable from my point of view.

In the end, it's your choice.

File: libs-scala/test-evidence/tag/src/main/scala/com/daml/test/evidence/tag/Security.scala
Comment (cbley-da) [reply libs-scala/test-evidence/tag/src/main/scala/com/daml/test/evidence/tag/Security.scala:122]: What exactly do you mean by "Avoid duplicate properties (Authenticity vs Authentication, Privacy vs Confidentiality)" -- you want me to revise all existing `SecurityTest`s and replace Authenticity with Authentication, Privacy with Confidentiality?

Comment (matthiasS-da) [reply libs-scala/test-evidence/tag/src/main/scala/com/daml/test/evidence/tag/Security.scala:122]: Remove either Authenticity or Authentication. Remove either Privacy or Confidentiality.

Up to you which one you drop. Feel free to drop the one that causes less migration effort.

Comment (cbley-da) [reply libs-scala/test-evidence/tag/src/main/scala/com/daml/test/evidence/tag/Security.scala:122]: Got it. Done.

Comment (matthiasS-da) [reply libs-scala/test-evidence/tag/src/main/scala/com/daml/test/evidence/tag/Security.scala:122]: > Remove either Authenticity or Authentication. Remove either Privacy or Confidentiality.

According to the diff, you haven't removed any of Authenticity, Authentication, Privacy, Confidentiality. So we end up with multiple properties for the same thing. This is exactly the situation I wanted to avoid in the first place.

To be frank, given this outcome I regret having pre-approved the PR.

Comment (matthiasS-da) [reply libs-scala/test-evidence/tag/src/main/scala/com/daml/test/evidence/tag/Security.scala:122]: @cbley-da

**Pull request merged**
